core/modules.rs

```diff
@@ -385,154 +384,112 @@ impl ModuleLoader for NoopModuleLoader {
-    maybe_load_callback: Option<ExtModuleLoaderCb>,
+    module_loader: Rc<dyn ModuleLoader>,
+    extensions: &[Extension],
+    module_map: &ModuleMap,
+    maybe_load_callback: Option<Rc<ExtModuleLoaderCb>>,
   ) -> Self {
-    let used_esm_sources: HashMap<String, bool> = extensions
-      .borrow()
-      .iter()
-      .flat_map(|e| e.get_esm_sources())
-      .flatten()
-      .map(|file_source| (file_source.specifier.to_string(), false))
-      .collect();
-
+    let mut resolutions = HashMap::new();
+    resolutions.extend(
+      extensions
+        .iter()
+        .flat_map(|e| e.get_esm_sources())
+        .flatten()
+        .map(|s| (s.specifier.to_string(), Some(s.clone()))),
+    );
+    resolutions.extend(
+      module_map
+        .names()
+        .iter()
+        .map(|n| (n.as_str().to_string(), None)),
+    );
     ExtModuleLoader {
-      module_loader: module_loader.unwrap_or_else(|| Rc::new(NoopModuleLoader)),
-      extensions,
-      ext_resolution_allowed: Default::default(),
-      used_esm_sources: RefCell::new(used_esm_sources),
+      module_loader,
       maybe_load_callback,
+      resolutions: RefCell::new(resolutions),
+      used_specifiers: Default::default(),
     }
   }
+}
 
-  pub fn resolve(
+impl ModuleLoader for ExtModuleLoader {
+  fn resolve(
     &self,
     specifier: &str,
     referrer: &str,
     kind: ResolutionKind,
   ) -> Result<ModuleSpecifier, Error> {
-    if specifier.starts_with("ext:") {
-      if !referrer.starts_with("ext:") && referrer != "."
-        || !*self.ext_resolution_allowed.borrow()
-      {
-        return Err(generic_error(
-          "Cannot load extension module from external code",
-        ));
-      }
+    if self.resolutions.borrow().contains_key(specifier) {
       return Ok(ModuleSpecifier::parse(specifier)?);
     }
     self.module_loader.resolve(specifier, referrer, kind)
   }
 
-  pub fn load(
+  fn load(
     &self,
-    module_specifier: &ModuleSpecifier,
+    specifier: &ModuleSpecifier,
     maybe_referrer: Option<&ModuleSpecifier>,
     is_dyn_import: bool,
   ) -> Pin<Box<ModuleSourceFuture>> {
-    if module_specifier.scheme() != "ext" {
-      return self.module_loader.load(
-        module_specifier,
-        maybe_referrer,
-        is_dyn_import,
-      );
-    }
-
-    let specifier = module_specifier.to_string();
-    let extensions = self.extensions.borrow();
-    let maybe_file_source = extensions
-      .iter()
-      .find_map(|e| e.find_esm(module_specifier.as_str()));
-
-    if let Some(file_source) = maybe_file_source {
-      {
-        let mut used_esm_sources = self.used_esm_sources.borrow_mut();
-        let used = used_esm_sources.get_mut(file_source.specifier).unwrap();
-        *used = true;
-      }
-
+    if let Some(source) = self.resolutions.borrow().get(specifier.as_str()) {
+      let source = source.as_ref().expect("modules missing a source are already in the module map and should not be loaded");
```

Comment:
Could you improve this message in `expect`? I can't really understand its meaning